app/src/main/java/com/islaharper/dawnofandroid/presentation/common/OneTap.kt

```diff
@@ -0,0 +1,133 @@
+package com.islaharper.dawnofandroid.presentation.common
+
+import android.app.Activity
+import android.util.Log
+import androidx.activity.compose.ManagedActivityResultLauncher
+import androidx.activity.compose.rememberLauncherForActivityResult
+import androidx.activity.result.ActivityResult
+import androidx.activity.result.IntentSenderRequest
+import androidx.activity.result.contract.ActivityResultContracts
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.ui.platform.LocalContext
+import com.google.android.gms.auth.api.identity.BeginSignInRequest
+import com.google.android.gms.auth.api.identity.Identity
+import com.google.android.gms.common.api.ApiException
+import com.google.android.gms.common.api.CommonStatusCodes
+import com.islaharper.dawnofandroid.util.Constants.CLIENT_ID
+
+@Composable
+fun StartActivityForResult(
+    key: Any,
+    onResultReceived: (String) -> Unit,
+    onDialogDismissed: () -> Unit,
+    launcher: (ManagedActivityResultLauncher<IntentSenderRequest, ActivityResult>) -> Unit,
+) {
+    val activity = LocalContext.current as Activity
+    val activityLauncher = rememberLauncherForActivityResult(
+        contract = ActivityResultContracts.StartIntentSenderForResult(),
+    ) { result ->
+        try {
+            if (result.resultCode == Activity.RESULT_OK) {
+                val oneTapClient = Identity.getSignInClient(activity)
+                val credentials = oneTapClient.getSignInCredentialFromIntent(result.data)
+                val tokenId = credentials.googleIdToken
+
+                if (tokenId != null) {
+                    onResultReceived(tokenId)
+                }
+            } else {
+                onDialogDismissed()
+            }
+        } catch (ex: ApiException) {
+            when (ex.statusCode) {
+                CommonStatusCodes.CANCELED -> {
+                    onDialogDismissed()
+                }
+
+                CommonStatusCodes.NETWORK_ERROR -> {
+                    onDialogDismissed()
+                }
+
+                else -> {
+                    onDialogDismissed()
+                }
+            }
+        }
+    }
+
+    LaunchedEffect(key1 = key) {
+        launcher(activityLauncher)
+    }
+}
+
+fun signIn(
```

Comment:
This seems quite tricky to design. When I look at the `gms.auth.api.identity.Identity` class, you can get a client using the `Activity` context  or just a regular `Context`. So maybe it is possible to arrive at a solution where you have an injectable Repository with access to `SignInClient` created using the application context. This repository could be a dependency for a ViewModel? This would promote testability of the code, since the ViewModel can be unit tested.

There would still be a place for some of the logic here, it would not be possible to move everything to the ViewModel. You would still need some way to transmit data from the ViewModel to the UI layer in order to invoke the `launchActivityForResult`s

What do you think?

Reply:
That sounds like a good way of moving forward. I have committed changes, hopefully it aligns with your idea.

Reply:
Refactored. Used this informative article to implement the changes:
https://medium.com/firebase-developers/how-to-authenticate-to-firebase-using-google-one-tap-in-jetpack-compose-60b30e621d0d

Reply:
Unfortunately I can't read that article because it is paywalled. I tried in normal browser and in private browsing session. Maybe you can attach a PDF printout to this PR?

Reply:
Added PDF to PR